### DISCUSSION TITLE
pypowerwall - memory usage

### DISCUSSION THREAD

**Discussion by youzer-name** [jasonacox/Powerwall-Dashboard#25] [Q&A] [answered]
I use Grafana and InfluxDB to monitor my Pi4's vital stats. Recently I noticed that my memory used was steadily climbing. When I checked the Task Manager in the Pi4 GUI I found that Python3 was using a lot of memory in the "VM-Size" column. After rebooting the Pi the total memory use dropped down a lot.

I've been watching the memory usage since my last reboot about 6 days ago. While pypowerwall isn't the only thing using more memory, restarting my pypowerwall container immediately dropped the total memory usage by about 11%. The memory use by Python3 dropped from over 400MB to < 20MB. Based on the numbers I saw before the reboot, I expect that if I had let it go longer, the Python3 memory utilization would have just continued to climb.

Is this normal or could there be a memory leak in pypowerwall or Python3?

This image shows the last 7 days. Towards the left you can see the before/after reboot. At the very right edge you can see the restart of the pypowerwall docker container.

![image](https://user-images.githubusercontent.com/57046429/166814844-a3235f3c-7703-41e2-b81f-eae4538342f6.png)

**Comment by jasonacox** [comment, marked as answer]
Hi @youzer-name !  Thanks for this observation...good find! 

It seems very compelling that there is a leak.  It should grow as it buffers API calls but that should level off and not continue to grow as your graph shows.  The pypowerwall proxy uses ThreadingHTTPServer.  I did some quick research and believe there are conditions with python 3.7+ with ThreadingHTTPServer that can cause this memory leak.  I have put in a recommended fix and I'm running some tests before releasing.  

Can you help by confirming the version of python as well as pypowerwall (see http://localhost:8675/version)?

I'm going to open an issue in https://github.com/jasonacox/pypowerwall for the proxy.

**Reply by jasonacox** [reply]
@youzer-name If you want to test the patch - Edit powerwall.yml and change this line:

```yml
# from
    pypowerwall:
        image: jasonacox/pypowerwall:latest

# to
    pypowerwall:
        image: jasonacox/pypowerwall:0.4.0t10
```

Delete and pull down the t10 proxy:

```bash
docker stop pypowerwall

# delete pypowerwall
docker rm pypowerwall
docker images | grep pypowerwall | awk '{print $3}' | xargs docker rmi -f

# rebuild
docker-compose -f powerwall.yml up -d
```

**Reply by jasonacox** [reply]
https://github.com/jasonacox/pypowerwall/issues/13

**Reply by jasonacox** [reply]
Update: The fix has been pushed to Dockerhub and is now available for upgrade (https://github.com/jasonacox/pypowerwall/issues/13). 

<img width="738" alt="image" src="https://user-images.githubusercontent.com/836718/167063917-5c68584b-02d4-4552-87ee-eef015d5e18e.png">

To upgrade, ensure your `powerwall.yml` has this (latest):
```yml
    pypowerwall:
        image: jasonacox/pypowerwall:latest
```

To pull down the upgrade:

```bash
docker stop pypowerwall

# delete pypowerwall
docker rm pypowerwall
docker images | grep pypowerwall | awk '{print $3}' | xargs docker rmi -f

# rebuild
docker-compose -f powerwall.yml up -d
```
